### PR TITLE
Adapt Java compilation to use the Problems API

### LINKED ISSUE #26529: Java compiler warnings and errors should be exposed via the problems API
We should receive, transform, and report Java compilation diagnostics using the Problems API.

### PR DESCRIPTION
**Pull request by hegyibalint**
Fixes #26529 and #27170

### CODE REVIEW

**Comment by hegyibalint** [comment]
@bot-gradle help

**Comment by hegyibalint** [comment]
@bot-gradle test PT

**Comment by hegyibalint** [comment]
@bot-gradle test APT

**Comment by hegyibalint** [comment]
@bot-gradle test RFM

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy [hunk 1]
```diff
@@ -14,18 +14,29 @@
...
 
+    // The `assertProblem` method will store visited files here, and `assertAllFilesVisited` will check it
+    List<String> assertedFileLocations = []
+
     def setup() {
+        enableProblemsApiCheck()
+
+        propertiesFile << """
+            # Feature flag as of 8.6 to enable the Problems API
+            org.gradle.compile.use-problems-api=true
```
Comment (donat) [inline platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:37]: Same as above. We'd need a feature flag when your changes would have a major impact by default.

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 2]
```diff
@@ -34,35 +33,56 @@
...
+            }
+
+            // We only set the location if we have a resource to point to
+            if (resourceName != null) {
+                spec.lineInFileLocation(resourceName, line, column, length);
+            }
+        });
 
         // We need to print the message to stderr as well, as it was the default behavior of the compiler
         System.err.println(message);
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:79]: This is the default behaviour when the diagnostic listener is null,right?

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 3]
```diff
@@ -34,35 +33,56 @@
...
         String message = diagnostic.getMessage(Locale.getDefault());
         String label = mapKindToLabel(diagnostic.getKind());
-        String resourceName = diagnostic.getSource() != null ? diagnostic.getSource().toUri().toString() : null;
-        Integer line = Math.toIntExact(diagnostic.getLineNumber());
-        Integer column = Math.toIntExact(diagnostic.getColumnNumber());
-        Integer length = Math.toIntExact(diagnostic.getEndPosition() - diagnostic.getStartPosition());
+
+        String resourceName = diagnostic.getSource() != null ? getPath(diagnostic.getSource()) : null;
+        int line = Math.toIntExact(diagnostic.getLineNumber());
+        int column = Math.toIntExact(diagnostic.getColumnNumber());
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:49]: ❓ The line and column values are 1-indexed, right?

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompileSpec.java [hunk 4]
```diff
@@ -70,4 +70,8 @@ public interface JavaCompileSpec extends JvmLanguageCompileSpec {
     default boolean annotationProcessingConfigured() {
         return !getAnnotationProcessorPath().isEmpty() && !getCompileOptions().getCompilerArgs().contains("-proc:none");
     }
+
+    void setProblemsApiReporting(boolean useProblemsApiReporting);
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompileSpec.java:74]: Same as above (or below, I don't see ordering in this view :) ).

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompiler.java [hunk 5]
```diff
@@ -45,34 +45,38 @@ public class JdkJavaCompiler implements Compiler<JavaCompileSpec>, Serializable
     private final DiagnosticToProblemListener diagnosticToProblemListener;
 
     @Inject
-    public JdkJavaCompiler(Factory<JavaCompiler> javaHomeBasedJavaCompilerFactory, Problems problems) {
+    public JdkJavaCompiler(Factory<JavaCompiler> javaHomeBasedJavaCompilerFactory, InternalProblems problems) {
```
Comment (donat) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompiler.java:48]: 💅 rename parameter to `problemsService`.

**Comment by hegyibalint** [comment]
@bot-gradle test APT

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJavaCompileSpec.java [hunk 6]
```diff
@@ -139,4 +140,15 @@ public List<File> getSourceRoots() {
     public void setSourcesRoots(List<File> sourcesRoots) {
         this.sourceRoots = sourcesRoots;
     }
+
+    @Override
+    public void setProblemsApiReporting(boolean useProblemsApiReporting) {
+        this.useProblemsApiReporting = useProblemsApiReporting;
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJavaCompileSpec.java:146]: > I think we should always emit the Problems API reports

I think we've agreed that the feature should belong behind a feature flag. 
This is part of that architecture, as I caught the property pretty low in the call chain (in the `JavaCompile` to be exact), and it needs this spec to get the message down to the compiler instance.

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy (hunk 1 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:37]: But isn't this a feature flag for you? 
Would you like to introduce a new `--enable-compilation-problems` flag?

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 3 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:49]: Checked (as it's undocumented), and they are

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 2 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:79]: Forgot about this line...
Technically _yes_, but this is not a nicely formatted message.

I think we should have this, until we figure out the rendering of the compilation problems.

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompileSpec.java (hunk 4 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompileSpec.java:74]: https://github.com/gradle/gradle/pull/27539/files#r1458699522

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJavaCompileSpec.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJavaCompileSpec.java:146]: But isn't this a feature flag for you? 
Would you like to introduce a new `--enable-compilation-problems` flag?

File: platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy (hunk 1 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileProblemsIntegrationTest.groovy:37]: https://github.com/gradle/gradle/pull/27539/files#r1452337507

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompiler.java (hunk 5 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompiler.java]: Renamed

**Comment by hegyibalint** [comment]
@bot-gradle test APT

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJavaCompileSpec.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DefaultJavaCompileSpec.java]: Discussed on the 📞

File: platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/BaseLocationTransformer.java [hunk 7]
```diff
@@ -36,10 +36,9 @@ public BaseLocationTransformer(BuildOperationAncestryTracker buildOperationAnces
     }
 
     @Nonnull
-    protected Optional<OperationIdentifier> getExecuteTask(Class<?> operationDetailsClass) {
+    protected Optional<OperationIdentifier> getExecuteTask(Class<?> operationDetailsClass, OperationIdentifier operationId) {
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/ide/problems/src/main/java/org/gradle/problems/internal/transformers/BaseLocationTransformer.java]: Renamed

File: platforms/jvm/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/GroovyCompilerFactory.java [hunk 8]
```diff
@@ -77,10 +77,10 @@ public Compiler<GroovyJavaJointCompileSpec> newCompiler(GroovyJavaJointCompileSp
     public static class DaemonSideCompiler implements Compiler<GroovyJavaJointCompileSpec> {
         private final ProjectLayout projectLayout;
         private final List<File> javaCompilerPlugins;
-        private final Problems problems;
+        private final InternalProblems problems;
 
         @Inject
-        public DaemonSideCompiler(ProjectLayout projectLayout, List<File> javaCompilerPlugins, Problems problems) {
+        public DaemonSideCompiler(ProjectLayout projectLayout, List<File> javaCompilerPlugins, InternalProblems problems) {
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/jvm/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/GroovyCompilerFactory.java]: Renamed

**Review by donat — APPROVED** [review]

**Comment by hegyibalint** [comment]
@bot-gradle cancel

**Comment by hegyibalint** [comment]
@bot-gradle merge

**Review by ghale — APPROVED** [review]
LGTM

**Comment by hegyibalint** [comment]
@bot-gradle merge

**Comment by cobexer** [comment]
@bot-gradle merge

File: platforms/core-execution/worker-processes/build.gradle.kts [hunk 9]
```diff
@@ -15,6 +15,7 @@ dependencies {
     api(project(":messaging"))
     api(project(":problems-api"))
     api(project(":process-services"))
+    api("com.google.code.findbugs:jsr305:3.0.2")
```
Comment (big-guy) [inline platforms/core-execution/worker-processes/build.gradle.kts:18]: This should be `libs.jsr305`

Comment (donat) [reply platforms/core-execution/worker-processes/build.gradle.kts:18]: https://github.com/gradle/gradle/pull/28087

**Pull request merged**